src/software/thunderscope/robot_communication.py

```diff
@@ -229,14 +245,20 @@ def __enter__(self):
             self.multicast_channel + "%" + self.interface, PRIMITIVE_PORT, True
         )
 
+        self.running = True
+
         self.send_estop_state_thread.start()
-        self.run_world_thread.start()
         self.run_primitive_set_thread.start()
 
+        if self.run_worlds:
+            self.run_world_thread.start()
+
         return self
 
     def __exit__(self):
         """Exit RobotCommunication context manager
 
         """
-        self.run_thread.join()
+        self.run_world_thread.join()
+        self.run_primitive_set_thread.join()
+        self.running = False
```

Comment:
ohh I was confusing thread yield and join, I assumed that join would auto stop the thread when the loop ends, so setting the boolean after the join would work. But it waits for the thread to end by stopping the current thread so that wouldn't work